ASoC: SOF: Intel: MTL: catch invalid IRQ IP ptr

yongzhi1 opened the pull request:
Add sanity check to catch 1FFFF upon resume to avoid page fault, when it happens very very rarelly.

plbossart commented on sound/soc/sof/intel/mtl.c:
```diff
@@ -83,6 +87,10 @@ static bool mtl_dsp_check_sdw_irq(struct snd_sof_dev *sdev)
 	hfintipptr = snd_sof_dsp_read(sdev, HDA_DSP_BAR, MTL_HFINTIPPTR) & MTL_HFINTIPPTR_PTR_MASK;
+	/* sanity check the irq ptr */
+	if (hfintipptr == MTL_HFINTIPPTR_PTR_MASK)
+		return false;
```
this shouldn't happen IMHO @yongzhi1. It's likely that we have a power-up issue somewhere and some parts of the IP are not up when we read that value.

@bardliao @RanderWang please chime in, you're probably the two contributors with the best understanding of the register layout and power dependencies.

yongzhi1 replied:
Thanks @plbossart , will talk offline with @bardliao @RanderWang on the failure scenario.

bardliao replied:
@plbossart we check all types of irq in hda_dsp_interrupt_thread(). Can any interrupt happens before DSP is powered up?
BTW, we have similar check in [hda_common_check_sdw_irq](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/intel/hda.c#L340) and [hda_dsp_check_ipc_irq](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/intel/hda-ipc.c#L366)

RanderWang replied:
It is very likely that we get a interrupt function call before dsp is powered up. MTL_HFINTIPPTR is also checked by mtl_enable_interrupts but why no error happen ?  Only reason is that mtl_enable_interrupts is not called before irq function accessing MTL_HFINTIPPTR since Linux interrupt framework invokes our registered irq function to process interrupt.

I recommend that we check power up state (a sw flag) in "hda_dsp_interrupt_thread(int irq, void *context)" and return early to avoid this issue. The check should be good for all platforms

plbossart replied:
> I recommend that we check power up state (a sw flag) in "hda_dsp_interrupt_thread(int irq, void *context)" and return early to avoid this issue.

can you clarify this suggestion @RanderWang? I am not following the proposal at all.

ranj063 replied:
@RanderWang I did some experiements with @yongzhi1 yesterday and it doesnt look like the DSP not being powered up will cause an exception.

RanderWang replied:
@ranj063  was mtl_enable_interrupts called ? MTL_HFINTIPPTR is also checked by mtl_enable_interrupts

RanderWang replied:
@plbossart  @ranj063 @bardliao  I do a further check that it may be related to our interrupt status check.
we have two registers for irq and only INSTS.GIS is checked.
#define SOF_HDA_INTCTL			0x20 
#define SOF_HDA_INTSTS			0x24

 According  to spec: even if the GIE bit (Global Interrupt Enable in INTCTL) is set to a 0 to disable processor interrupt generation, the GIS bit may still be set to indicate that stream is requesting service.

A recommend check should be like

````c
static irqreturn_t hda_dsp_interrupt_handler(int irq, void *context)
{
	struct snd_sof_dev *sdev = context;

	/*
	 * Get global interrupt status. It includes all hardware interrupt
	 * sources in the Intel HD Audio controller.
	 */
-	if (snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR, SOF_HDA_INTSTS) &
-	    SOF_HDA_INTSTS_GIS) {
+      intsts = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR, SOF_HDA_INTSTS);
+      intctl = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR, SOF_HDA_INTCTL);

+      /* both GIS and GIE are in bit 31, so check GIE & GIS to make sure interrupt is real one or not */
+      intval = (intsts & intctl);
+      if (intval & SOF_HDA_INTSTS_GIS)
		/* disable GIE interrupt */
		snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR,
					SOF_HDA_INTCTL,
					SOF_HDA_INT_GLOBAL_EN,
					0);

		return IRQ_WAKE_THREAD;
	}
````

yongzhi1 replied:
Ack, will test and update.

ujfalusi replied:
The page fault is caused by the register read when the DSP is off, no?

yongzhi1 commented:
Close now, will revisit if the original issue reoccurs.

yongzhi1 commented:
It's ironic, got pinged today with another incidence of the same issue, can we review the patch again?

```
[   23.447007] BUG: unable to handle page fault for address: ffffa8c24c40001f
[   23.454706] #PF: supervisor read access in kernel mode
[   23.460461] #PF: error_code(0x0000) - not-present page
[   23.466214] PGD 101000067 P4D 101000067 PUD 101201067 PMD 1168d2067 PTE 0
[   23.473817] Oops: 0000 [#1] PREEMPT SMP NOPTI
[   23.478694] CPU: 8 PID: 3028 Comm: irq/193-AudioDS Tainted: G     U             6.1.28-ga10bfb5d86a4 #1 ec5ee46f96743fee3e488134312748dcafcb4236
[   23.493193] Hardware name: Google Rex/Rex, BIOS Google_Rex.15456.0.0 05/18/2023
[   23.501368] RIP: 0010:mtl_dsp_check_ipc_irq+0x8e/0xef [snd_sof_intel_hda_common]
[   23.509674] Code: 48 8b 49 20 48 8b 89 98 00 00 00 4c 8b 59 58 48 8d 34 10 48 83 c6 20 4d 85 db 74 0e 48 89 df 41 ff d3 0f 1f 00 41 89 c6 eb 03 <44> 8b 36 66 90 41 83 fe ff 0f 95 c0 41 20 c6 44 89 f0 5b 41 5e 5d
[   23.530678] RSP: 0018:ffffa8c24a8f7e28 EFLAGS: 00010246
[   23.536527] RAX: 00000000001fffff RBX: ffffa063c4304828 RCX: ffffffffc0b93b90
[   23.544509] RDX: ffffa8c24c200000 RSI: ffffa8c24c40001f RDI: ffffa063c4304828
[   23.552494] RBP: ffffa8c24a8f7e38 R08: 00000005758c0f97 R09: 0000000000000001
[   23.560479] R10: 0000000000000008 R11: 0000000000000000 R12: ffffffffb63c0b64
[   23.568457] R13: ffffa063c43310c0 R14: 00000000000000c1 R15: ffffa063c4301028
[   23.576440] FS:  0000000000000000(0000) GS:ffffa06adfe00000(0000) knlGS:0000000000000000
[   23.585491] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[   23.591919] CR2: ffffa8c24c40001f CR3: 00000006f140c002 CR4: 0000000000770ee0
[   23.599899] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[   23.607881] DR3: 0000000000000000 DR6: 00000000ffff07f0 DR7: 0000000000000400
[   23.615856] PKRU: 55555554
[   23.618891] Call Trace:
[   23.621630]  <TASK>
[   23.623986]  hda_dsp_interrupt_thread+0x71/0x392 [snd_sof_intel_hda_common 4d8936f2cf1a2a3133519cb5419af2dc1256ed91]
[   23.635777]  ? irq_forced_thread_fn+0x61/0x61
[   23.640652]  irq_thread_fn+0x22/0x4f
[   23.644645]  irq_thread+0x104/0x1b7
[   23.648544]  ? irq_forced_thread_fn+0x61/0x61
[   23.653421]  ? irq_thread_fn+0x4f/0x4f
[   23.657620]  kthread+0xde/0xf7
[   23.661041]  ? irq_forced_secondary_handler+0x22/0x22
[   23.666696]  ? kthread_blkcg+0x32/0x32
[   23.670887]  ret_from_fork+0x1f/0x30
[   23.674891]  </TASK>
[   23.677331] Modules linked in: pps_core(+) vsock(+) rndis_wlan rndis_host cdc_ether usbnet mii uinput 8021q ccm iio_trig_hrtimer snd_soc_sof_rt5682 industrialio_sw_trigger industrialio_configfs snd_soc_hdac_hdmi snd_soc_intel_hda_dsp_common snd_soc_intel_sof_realtek_common snd_soc_intel_sof_maxim_common snd_soc_rt5682 snd_sof_probes snd_soc_rl6231 snd_soc_dmic btusb btrtl veth btintel btmtk btbcm snd_hda_codec_hdmi intel_ipu6_psys snd_sof_pci_intel_mtl snd_sof_intel_hda_common snd_sof_intel_hda soundwire_intel soundwire_generic_allocation soundwire_cadence snd_sof_pci snd_sof_xtensa_dsp snd_soc_hdac_hda snd_soc_acpi_intel_match snd_soc_acpi snd_hda_ext_core snd_sof intel_ipu6_isys snd_sof_utils xt_cgroup soundwire_bus videobuf2_dma_contig snd_intel_dspcfg videobuf2_v4l2 snd_intel_sdw_acpi videobuf2_memops snd_hda_codec videobuf2_common snd_hwdep intel_pmc_mux ov13b10 snd_hda_core snd_soc_max98357a hi556 xt_MASQUERADE acpi_als cros_ec_typec roles v4l2_fwnode dw9714 snd_soc_rt5682s
[   23.677403]  v4l2_async at24 typec intel_ipu6 ip6table_nat fuse dm_integrity async_xor xor async_tx iwlmvm iwl7000_mac80211 bluetooth iio_trig_sysfs iwlwifi cros_ec_lid_angle cros_ec_light_prox cros_ec_sensors cros_ec_sensors_core ecdh_generic industrialio_triggered_buffer ecc kfifo_buf cfg80211 industrialio cros_ec_sensorhub lzo_rle lzo_compress zram zsmalloc joydev
[   23.815245] gsmi: Log Shutdown Reason 0x03
[   23.819835] CR2: ffffa8c24c40001f
[   23.823549] ---[ end trace 0000000000000000 ]---
[   23.830447] RIP: 0010:mtl_dsp_check_ipc_irq+0x8e/0xef [snd_sof_intel_hda_common]
[   23.838765] Code: 48 8b 49 20 48 8b 89 98 00 00 00 4c 8b 59 58 48 8d 34 10 48 83 c6 20 4d 85 db 74 0e 48 89 df 41 ff d3 0f 1f 00 41 89 c6 eb 03 <44> 8b 36 66 90 41 83 fe ff 0f 95 c0 41 20 c6 44 89 f0 5b 41 5e 5d
[   23.859762] RSP: 0018:ffffa8c24a8f7e28 EFLAGS: 00010246
[   23.865599] RAX: 00000000001fffff RBX: ffffa063c4304828 RCX: ffffffffc0b93b90
[   23.873579] RDX: ffffa8c24c200000 RSI: ffffa8c24c40001f RDI: ffffa063c4304828
[   23.881551] RBP: ffffa8c24a8f7e38 R08: 00000005758c0f97 R09: 0000000000000001
[   23.889526] R10: 0000000000000008 R11: 0000000000000000 R12: ffffffffb63c0b64
[   23.897499] R13: ffffa063c43310c0 R14: 00000000000000c1 R15: ffffa063c4301028
[   23.905472] FS:  0000000000000000(0000) GS:ffffa06adfe00000(0000) knlGS:0000000000000000
[   23.914522] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[   23.920941] CR2: ffffa8c24c40001f CR3: 00000006f140c002 CR4: 0000000000770ee0
[   23.928913] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[   23.936884] DR3: 0000000000000000 DR6: 00000000ffff07f0 DR7: 0000000000000400
[   23.944856] PKRU: 55555554
[   23.947875] Kernel panic - not syncing: Fatal exception
[   23.953843] Kernel Offset: 0x35200000 from 0xffffffff81000000 (relocation range: 0xffffffff80000000-0xffffffffbfffffff)
```

yongzhi1 commented:
Thanks @bardliao @RanderWang for the review, let me include it to our engineering build first and see if it helps :)

plbossart commented:
@yongzhi1 any update on this PR, is this needed?

yongzhi1 commented:
Hi, @plbossart , I have not heard any feedback on internal validation , let me close and keep an eye on it.

yongzhi1 committed:
ASoC: SOF: Intel: hda: add a safety check in hda_dsp_interrupt_handler

Check INTCTL GIE (Global Interrupt Enable) bit before handling the IRQ
to avoid any potential races at suspend & resume.

Signed-off-by: Yong Zhi <yong.zhi@intel.com>

yongzhi1 commented:
Re-open in light of the recent surge of https://github.com/thesofproject/sof/issues/7866.

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1076,6 +1076,13 @@ static irqreturn_t hda_dsp_interrupt_handler(int irq, void *context)
+		pr_warn_once("%s: spurious interrupt\n", __func__);
+		return IRQ_NONE;
+	}
+
```
if you are reading a register while the DSP is off, isn't this going to return 0xFFFFFFFF and that means this test will not help?

ujfalusi replied:
It will likely 'just' cause page fault at different line.

yongzhi1 replied:
The page fault was not caused by DSP being off as commented here:
https://github.com/thesofproject/linux/pull/4344#discussion_r1190301714

The call trace quoted in this PR as well as the following one pointed to the same place:
https://github.com/thesofproject/sof/files/12404629/kernel.20230815.205835.87076.0.txt

The patch is trying to avoid the page fault only, it's not a root-cause fix
```
<4>[36962.878771] RAX: 00000000001fffff RBX: ffff9ce686a3c028 RCX: ffffffffc07fdb9
```

The attempt is quite clear in the original version, but updated based reviews.

yongzhi1 replied:
DId some experiments on my board, whether SOF_HDA_INT_GLOBAL_EN set or not, snd_sof_dsp_read(sdev, HDA_DSP_BAR, MTL_HFINTIPPTR) always returns 1800, so using SOF_HDA_INT_GLOBAL_EN wouldn't catch the case. The only explanation is that it tried to read MTL_HFINTIPPTR when HDA_DSP_BAR is gated.

RanderWang replied:
How did you test SOF_HDA_INT_GLOBAL_EN ? SOF_HDA_INT_GLOBAL_EN indicates internal hw is not ready, how do you build this state ?

yongzhi1 replied:
Hi, @RanderWang, I checked MTL_HFINTIPPTR value here:
https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/intel/hda.c#L1216

So my conclusion is that SOF_HDA_INT_GLOBAL_EN is 0 does not necessitate the condition for incorrect MTL_HFINTIPPTR value.

Also since @ujfalusi is working on a better solution for this issue, let me close this one.

plbossart reviewed (commented):
I don't get what this fixes @yongzhi1
